### ISSUE TITLE
ER: New Developers with Mac processors are having problems running docker image

### ISSUE THREAD

**Issue by roslynwythe** [hackforla/website#5644]
### Dependencies (Issues created from this ER)
- [ ] #6266

### Emergent Requirement - Problem
A new developer with a Mac running M1 is unable to run docker using the hfla ghpages docker image

### Issue you discovered this emergent requirement in
- Slack message: https://hackforla.slack.com/archives/C4UM52W93/p1694561505275069?thread_ts=1694556538.666779&cid=C4UM52W93 also copied to this comment below: https://github.com/hackforla/website/issues/5644#issuecomment-1807752887

### Date discovered 
9/12

### Did you have to do something temporarily
- [ ] YES
- [ ] NO

### Who was involved
@sophia-bui @fyliu @homeroochoa47 

### What happens if this is not addressed
Some Mac users will be unable to run docker

### Recommended Action Items
- [ ] Make a new issue
- [ ] Discuss with team
- [ ] Let a Team Lead know

### Potential solutions [draft]
We may do a variety of solutions, not just one.  We should make issues (if feasible) for each.
#### Solution 1 (preferred)
- Review the docker image Hack for LA created https://github.com/hackforla/ops/wiki/@danielridgebot-Ops-Wiki and see what we need to do to update it.   
   - Make sure its a multi-platform image (if its not, improve it)
#### Solution 2
- Provide instructions in CONTRIBUTING.md for devs to build their own image using the Dockerfile
#### Solution 3
- Provide instructions in CONTRIBUTING.md for devs to use the Rosetta 2 emulator

**Comment by roslynwythe** [comment]
Sophia Bui
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694556538666779)
Hi,
I’m trying to set up my dev environment and ran into this issue yesterday. I’m currently on this [link](https://github.com/hackforla/website/blob/gh-pages/CONTRIBUTING.md#part-1-setting-up-the-development-environment) (on 1.6 Dev setup (6): Build and serve the website locally 1.6.a Starting Docker). I have Docker set up and have used it a few months ago. The instructions say to just have the application running and run the command “docker-compose up.” I did that, but I’m not getting the expected terminal output. I have a mac M1 and the output/error message I received is below (1. platform error 2. jekyll config file):
 ! hfla_site The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Attaching to hfla_site
hfla_site  | ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux-musl]
hfla_site  | Configuration file: none
hfla_site  |            Cleaner: Nothing to do for /srv/jekyll/_site.
hfla_site  |            Cleaner: Nothing to do for /srv/jekyll/.jekyll-metadata.
hfla_site  |            Cleaner: Nothing to do for .sass-cache.
hfla_site  | ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux-musl]
hfla_site  |              Fatal: The configuration file '_config.yml' could not be found.
hfla_site  | /usr/gem/gems/jekyll-3.9.2/lib/jekyll/configuration.rb:191:in `rescue in read_config_file': The Configuration file '_config.yml' could not be found. (LoadError)
hfla_site  | 	from /usr/gem/gems/jekyll-3.9.2/lib/jekyll/configuration.rb:179:in `read_config_file'
hfla_site  | 	from /usr/gem/gems/jekyll-3.9.2/lib/jekyll/configuration.rb:207:in `block in read_config_files'
28 replies


Fang
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694561505275069?thread_ts=1694556538.666779&cid=C4UM52W93)
Yes, this is a problem and I’m surprised it’s only surfacing now. The ghpages docker image only works for 64bit Intel. You should file a bug and ask that a multi platform image be built. Or at least one for arm 64 for m1
:+1:
1



Fang
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694562189022609?thread_ts=1694556538.666779&cid=C4UM52W93)
You can also build your own image using the Dockerfile in the meantime. (edited) 


Sophia Bui
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694563418677099?thread_ts=1694556538.666779&cid=C4UM52W93)
Thank you!


Sophia Bui
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694563782325259?thread_ts=1694556538.666779&cid=C4UM52W93)
I might go to office hours for this if I’m having trouble building the image


Fang
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694563865072839?thread_ts=1694556538.666779&cid=C4UM52W93)
It’s the hackforla/ghpages-docker repo. That’s where you need to go to get the instructions for the image.
:heart:
1



Homero Ochoa
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694565384475999?thread_ts=1694556538.666779&cid=C4UM52W93)
Hi 
[@Sophia Bui](https://hackforla.slack.com/team/U05Q5GAP3RQ)
, out of curiosity did you install Rosetta 2 when installing docker? I have an M2 mac  but haven't had this issue, and I believe Rosetta 2 should allow you to run the amd64 platform on a mac. [Here's](https://stackoverflow.com/questions/76848446/running-docker-containers-on-mac-apple-silicon-m2-platform-compatibility-and-ex) a similar issue from Stack Overflow
Stack OverflowStack Overflow
[Running Docker Containers on Mac Apple Silicon M2: Platform Compatibility and Execution Errors (linux/arm64)](https://stackoverflow.com/questions/76848446/running-docker-containers-on-mac-apple-silicon-m2-platform-compatibility-and-ex)
I'm a newbie to both macOS and Docker. I'm running into a series of roadblocks while trying to launch my project using Docker on Apple Silicon M2. I think that the issues i'm facing are related to ...
:+1:
1



Fang
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694565493185859?thread_ts=1694556538.666779&cid=C4UM52W93)
I think just cloning it and running docker compose on it will build it


Fang
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694567297403779?thread_ts=1694556538.666779&cid=C4UM52W93)
Right, Rosetta can emulate and allow you to run the image. It’s the simpler solution. Can one of you suggest that this be added to the documentation? Maybe then work on that issue.
:raised_hands:
1



Sophia Bui
  [2 months ago](https://hackforla.slack.com/archives/C4UM52W93/p1694573123230799?thread_ts=1694556538.666779&cid=C4UM52W93)
I used Docker in the past with another opensource and did have an issue being an M1. I believe I installed Rosetta before. I’ll see and also ask DevOps. I’ll document it if I figure it out, appreciate all the help!
:raised_hands:
1



Roslyn Wythe
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696227718759159?thread_ts=1694556538.666779&cid=C4UM52W93)
Thank you 
[@Fang](https://hackforla.slack.com/team/UCE01B4AX)
 
[@Sophia Bui](https://hackforla.slack.com/team/U05Q5GAP3RQ)
 
[@Homero Ochoa](https://hackforla.slack.com/team/U05K4RXFZ9P)
 for your ideas.  I wrote an ER https://github.com/hackforla/website/issues/5644 describing this situation.  Would any of you be interested in writing and/or working on the issue?


<https://github.com/[hackforla/website](https://github.com/hackforla/website)|hackforla/website>hackforla/website | Oct 1st | Added by [GitHub](https://hackforla.slack.com/services/B020G4XAAH5)


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696235757679199?thread_ts=1694556538.666779&cid=C4UM52W93)
[@Sophia Bui](https://hackforla.slack.com/team/U05Q5GAP3RQ)
 i dont think this is an m1/arm issue, since it looks like the container still manages to run. in the error it says _config.yml couldnt be found, do you have a _config.yml?


Sophia Bui
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696289240669899?thread_ts=1694556538.666779&cid=C4UM52W93)
That is great that there is an issue for this. I’m not interested right now since the issue/solution is not clear, I moved to a different team so I could get started on contributing.
Hi 
[@jason wong](https://hackforla.slack.com/team/U05TN1JQ0R5)
 — regarding the config file, I can’t say 100% yes that file was there since I don’t have a copy on my local machine anymore, however I believe it was there since I was working on the yml file… ? If the config file is on the repo, I cloned it exactly as it was and followed instructions. I tried to build my own image and tried to edit the config file to add in multiple platforms but it was not working out. I did everything stated on the docs on onboarding/setting up my environment, didn’t change or delete files except trying to add on a line to include multiple platforms on my local machine. I wouldn’t think to add another yml file?


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696289436168869?thread_ts=1694556538.666779&cid=C4UM52W93)
oh interesting, then im really not sure now then since it just ran on my mac


Sophia Bui
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696357108903749?thread_ts=1694556538.666779&cid=C4UM52W93)
[@jason wong](https://hackforla.slack.com/team/U05TN1JQ0R5)
  do you have an M1? It could be an M1 issue since there’s been many issues online with docker :woman-shrugging:


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696357134852039?thread_ts=1694556538.666779&cid=C4UM52W93)
yes i do


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696357176743359?thread_ts=1694556538.666779&cid=C4UM52W93)
❯ docker-compose up
[+] Running 1/0
 ⠿ Container hfla_site  Running                                                                                                         0.0s
Attaching to hfla_site
hfla_site  | Configuration file: /srv/jekyll/_config.yml
hfla_site  |            Cleaner: Removing /srv/jekyll/_site...
hfla_site  |            Cleaner: Removing ./.jekyll-metadata...
hfla_site  |            Cleaner: Nothing to do for .sass-cache.
hfla_site  | ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-linux-musl]
hfla_site  | Configuration file: _config.yml
hfla_site  | Configuration file: _config.docker.yml
hfla_site  |             Source: .
hfla_site  |        Destination: /srv/jekyll/_site
hfla_site  |  Incremental build: enabled
hfla_site  |       Generating...
hfla_site  |                     done in 133.182 seconds.
hfla_site  |  Auto-regeneration: enabled for '.'
hfla_site  | LiveReload address: http://0.0.0.0:35729/
hfla_site  |     Server address: http://0.0.0.0:4000/
hfla_site  |   Server running... press ctrl-c to stop.
hfla_site  |         LiveReload: Browser connected
^CGracefully stopping... (press Ctrl+C again to force)
[+] Running 1/1
 ⠿ Container hfla_site  Stopped                                                                                                         0.4s
canceled
this is the output i get on mine


Sophia Bui
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696357523539549?thread_ts=1694556538.666779&cid=C4UM52W93)
:flushed: Weird… I updated everything too like docker, Rosetta etc. Since no one else had this issue, maybe 
[@Roslyn Wythe](https://hackforla.slack.com/team/U046PD8UT55)
 can just delete the issue then. I’ll be working on other projects anyway


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696357771736389?thread_ts=1694556538.666779&cid=C4UM52W93)
I don’t have an apple silicon machine to test with if we made it build for arm too. But it seems like a really simple thing to make it happen that someone should do it anyway. Just a few lines in the ci config file.


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696359382463219?thread_ts=1694556538.666779&cid=C4UM52W93)
[@Roslyn Wythe](https://hackforla.slack.com/team/U046PD8UT55)
 do you know who maintains the ghpages docker image?
https://github.com/hackforla/ghpages-docker
hackforla/ghpages-docker
Docker file to build image pegged to GitHub Pages dependencies
Language
Shell
Last updated
a year ago
Added by [GitHub](https://hackforla.slack.com/services/B020G4XAAH5)


Roslyn Wythe
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696359745362299?thread_ts=1694556538.666779&cid=C4UM52W93)
No unfortunately I don't know who maintains the ghpages docker image, but I was hoping someone with sufficient expertise would pick up the issue.  I was hoping to ask Bonnie about it at our last meeting but there was not enough time.   
[@Fang](https://hackforla.slack.com/team/UCE01B4AX)
 do you know of any current members with this expertise who might advise us?


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696360912470549?thread_ts=1694556538.666779&cid=C4UM52W93)
something that may work as an alternative is to have users build their own version of the docker image in their machines platform
image.png
 
image.png


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696360946607469?thread_ts=1694556538.666779&cid=C4UM52W93)
i dont know if it solves sophia's issue in particular but it does build a native version of the image


Roslyn Wythe
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696360949641119?thread_ts=1694556538.666779&cid=C4UM52W93)
Yes @Jason.  Perhaps I should write an issue for someone to write up those instructions


Roslyn Wythe
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696360989004149?thread_ts=1694556538.666779&cid=C4UM52W93)
for  a dev to build their own version of the docker image


jason wong
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696361025644049?thread_ts=1694556538.666779&cid=C4UM52W93)
it'd just mean devs would have to run a one time docker-compose build (edited) 


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696361041440969?thread_ts=1694556538.666779&cid=C4UM52W93)
Looks good 
[@jason wong](https://hackforla.slack.com/team/U05TN1JQ0R5)
 . Eric Vennemeyer did the docker image. 


Roslyn Wythe
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696361086232879?thread_ts=1694556538.666779&cid=C4UM52W93)
[@jason wong](https://hackforla.slack.com/team/U05TN1JQ0R5)
 if its that simple, I'll write a complexity small issue to add that instruction to CONTRIBUTING.md


Fang
  [1 month ago](https://hackforla.slack.com/archives/C4UM52W93/p1696361090285779?thread_ts=1694556538.666779&cid=C4UM52W93)
Or have someone update that ghpages-docker repo to build the arm image. Either way works. (edited)

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty Please approve for issue making (should I have applied `ready for prioritization` instead?)

**Comment by ExperimentsInHonesty** [comment]
@roslynwythe For additional context, Eric V. Created the docker image, and the bot that is supposed to maintain it. It looks like it has stopped running.  Here is the WIKI file on ops that details where all the files are and how the update bot works. So that someone can start the journey to see why the updates stopped happening and what next steps are to get the docker image up to date and make sure the bot starts running again.

I have added updating this to the list of solutions above.

I am ready to prioritize this if you sign off on the way I have restructured the potential solutions.

I changed it from 
> ### Potential solutions [draft]
> - Developa multi-platform image
> - Provide instructions in CONTRIBUTING.md for devs to build their own image using the Dockerfile
> - Provide instructions on using Rosetta 2 emulator

to 

> ### Potential solutions [draft]
> We may do a variety of solutions, not just one. We should make issues (if feasible) for each.
> 
> #### Solution 1 (preferred)
> -Review the docker image Hack for LA created https://github.com/hackforla/ops/wiki/@danielridgebot-Ops-Wiki and see what we need to do to update it.
>    - Make sure its a multi-platform image (if its not, improve it)
> #### Solution 2
> - Provide instructions in CONTRIBUTING.md for devs to build their own image using the Dockerfile
> #### Solution 3
> - Provide instructions in CONTRIBUTING.md for devs to use the Rosetta 2 emulator

**Comment by jaasonw** [comment]
the bot likely stopped running because

> Warning: To prevent unnecessary workflow runs, scheduled workflows may be disabled automatically. When a public repository is forked, scheduled workflows are disabled by default. In a public repository, scheduled workflows are automatically disabled when no repository activity has occurred in 60 days.

https://docs.github.com/en/actions/using-workflows/disabling-and-enabling-a-workflow

**Comment by roslynwythe** [comment]
@ExperimentsInHonesty Yes that looks good.  It would be great to have all three options available.

**Comment by JessicaLucindaCheng** [comment]
@jaasonw 
> the bot likely stopped running because
> 
> > Warning: To prevent unnecessary workflow runs, scheduled workflows may be disabled automatically. When a public repository is forked, scheduled workflows are disabled by default. In a public repository, scheduled workflows are automatically disabled when no repository activity has occurred in 60 days.
> 
> https://docs.github.com/en/actions/using-workflows/disabling-and-enabling-a-workflow

Great catch! :smiley: I think a possible solution would be to have a GHA (GitHub Action) in the [hackforla/website repo](https://github.com/hackforla/website) run every 60 days to update a file in the [danielridgebot/check-ghpages-versions repo](https://github.com/danielridgebot/check-ghpages-versions). This file could just have something like "Last update" and then the date.

**Comment by roslynwythe** [comment]
There is a solution to the problem of the disabled workflow - "Keep-Alive Workflow", see https://github.com/marketplace/actions/keepalive-workflow

**Comment by roslynwythe** [comment]
- @ExperimentsInHonesty I suggest changes to the structure of "Potential Solutions", mainly because I want to separate the multi-platform enhancement (in the [hackforla/ghpages-docker](https://github.com/hackforla/ghpages-docker/blob/main/.github/workflows/build-and-push-to-docker-hub.yml) repo) from the investigation to the [danielridgebot/check-ghpages-versions](https://github.com/danielridgebot/check-ghpages-versions) repo, which is already the subject of an open issue #5897

Thanks to @jaasonw for information regarding the multi-platform image

### Potential Solutions

We should make issues (if feasible) as follows:
- [ ] Create issue for updating [hackforla/ghpages-docker](https://github.com/hackforla/ghpages-docker/blob/main/.github/workflows/build-and-push-to-docker-hub.yml) for multi-platform use
   - [ ]  see [GHA: Docker Setup Buildx](https://github.com/marketplace/actions/docker-setup-buildx) 
- [ ] Create issue for providing instructions in CONTRIBUTING.md for devs to build their own image using the Dockerfile
- [ ] Create issue for providing instructions in CONTRIBUTING.md for devs to use the Rosetta 2 emulator

**Comment by fyliu** [comment]
In case this is useful, here's [how I used Docker Buildx action in a similar project](https://github.com/hackforla/docker-mkdocs/blob/main/.github/workflows/build-image.yml).
[Here are the resulting images](https://hub.docker.com/repository/docker/hackforlaops/mkdocs/tags?page=1&ordering=last_updated).

**Comment by ExperimentsInHonesty** [comment]
LGTM

**Comment by roslynwythe** [comment]
- See #6266

**Comment by roslynwythe** [comment]
@fyliu Hi Fang, I'm writing the issues now to keep the Daniel Ridge bot alive and to update the Docker file so it supports multiple platforms.  I'm looking at the Docker file documentation https://docs.docker.com/build/building/multi-platform/ and I need to select the specific architecture.  I know mac M1 is arm-based but which specific architecture should we specify?  And should we support additional architectures, aside from x86_64 ?

Thanks for your expertise, we really appreciate it!

**Comment by fyliu** [comment]
In the repo I worked on above, I only chose the the most common platforms arm64 (Mac m1 and newer) and amd64 (intel Mac and most PCs). I was working on it at the time this issue came up and decided to add the support because of this.

There might be other useful platforms like Chrome/netbook or surface tablets, and raspberrypi 4 and above. I don't know if they're some different arm platforms. I kind of decide to leave it as an example in the code and someone can add more when needed.
